### ISSUE TITLE
problem in the simclr loss

### ISSUE THREAD

**Issue by drasros** [mpatacchiola/self-supervised-relational-reasoning#3]
Hi!
Thanks for this repo which is useful to us. 
I wanted to let you know that there is a problem in your implementation of the simclr loss. (Model.methods.simclr.Model.return_loss_fn)
the return value should be torch.mean(-torch.log(x)) instead of -torch.log(x.mean()). 
Kind regards,
A

**Comment by mpatacchiola** [comment]
Hi @drasros 

I am glad you found the repository to be useful.

Regarding the SimCLR loss, back in the day when I implemented the code, I have found online implementations that used both `torch.mean(-torch.log(x))` and `-torch.log(x.mean())`. I have tried both of them and in practical terms there is little or no difference in the results. However, in some conditions the form `-torch.log(x.mean())` seems to be better. For instance, I have run SimCLR with ResNet32 backbone on CIFAR-10 and CIFAR-100 with both losses. The average accuracy on three seeds is the following:

- `-torch.log(x.mean())` -> 77.02% (CIFAR-10); and 42.13% (CIFAR-100)
- `torch.mean(-torch.log(x))` -> 76.05% (CIFAR-10); and 42.19% (CIFAR-100)

As you can see, while on CIFAR-100 both losses give approximately the same accuracy, on CIFAR-10 there is a drop of ~1% when using the form you suggested. Anyhow, I will add a comment in the SimCLR code showing the alternative loss, so that the user can choose one or the other. 

Thank you.

**Comment by mpatacchiola** [comment]
I have included a comment about this in the `simclr.py` script.

**Comment by drasros** [comment]
Hi @mpatacchiola , sorry for the late reply and thank you for the details about your choice ;)
